alephium/lib/deployment.ts

```diff
@@ -0,0 +1,284 @@
+    }
+    const content = JSON.stringify(json, null, 2)
+    return fsPromises.writeFile(filepath, content)
+  }
+
+  static async from(filepath: string): Promise<Deployments | undefined> {
+    if (!fs.existsSync(filepath)) {
+      return undefined
+    }
+    const content = await fsPromises.readFile(filepath)
+    const json = JSON.parse(content.toString())
+    const lastFailedStep = json.lastFailedStep as number
+    const deployContractResults = new Map(Object.entries<DeployContractResult>(json.deployContractResults))
+    const runScriptResults = new Map(Object.entries<RunScriptResult>(json.runScriptResults))
+    return new Deployments(lastFailedStep, deployContractResults, runScriptResults)
+  }
+}
+
+export interface Deployer {
+  provider: NodeProvider
+  account: Account
+
+  deployContract(contract: Contract, params: DeployContractParams): Promise<DeployContractResult>
+  runScript(script: Script, params: RunScriptParams): Promise<RunScriptResult>
+
+  getDeployContractResult(typeId: string): DeployContractResult
+  getRunScriptResult(typeId: string): RunScriptResult
+}
+
+export type DeployFunction = (deployer: Deployer, networkType: NetworkType) => Promise<void>
+
+export class PrivateKeySigner extends SignerWithNodeProvider {
+  private static ec = new EC('secp256k1')
+
+  private readonly privateKey: string
+  private readonly publicKey: string
+  private readonly address: string
+  private readonly group: number
+
+  constructor(privateKey: string, alwaysSubmitTx = true) {
+    super(alwaysSubmitTx)
+    this.privateKey = privateKey
+    this.publicKey = publicKeyFromPrivateKey(privateKey)
+    this.address = addressFromPublicKey(this.publicKey)
+    this.group = groupOfAddress(this.address)
+  }
+
+  static from(mnemonic: string): PrivateKeySigner {
+    const wallet = getWalletFromMnemonic(mnemonic)
+    return new PrivateKeySigner(wallet.privateKey)
+  }
+
+  getAccountSync(): Account {
+    return { address: this.address, publicKey: this.publicKey, group: this.group }
+  }
+
+  async getAccounts(): Promise<Account[]> {
+    return [this.getAccountSync()]
+  }
+
+  async signRaw(signerAddress: string, hexString: string): Promise<string> {
+    if (signerAddress !== this.address) {
+      throw Error('Unmatched signer address')
+    }
+
+    const key = PrivateKeySigner.ec.keyFromPrivate(this.privateKey)
+    const signature = key.sign(hexString)
+    return signatureEncode(signature)
+  }
+}
+
+function createDeployer(
+  network: Network,
+  deployContractResults: Map<string, DeployContractResult>,
+  runScriptResults: Map<string, RunScriptResult>
+): Deployer {
+  const signer = PrivateKeySigner.from(network.mnemonic)
+  const account = signer.getAccountSync()
+
+  const deployContract = async (contract: Contract, params: DeployContractParams): Promise<DeployContractResult> => {
+    const result = await contract.transactionForDeployment(signer, params)
+    await signer.submitTransaction(result.unsignedTx, result.txId)
```

Comment:
Do you think it makes sense to deploy only when contract code is changed? We probably can make the assumption that the deployment steps are supposed to be idempotent, therefore can be run multiple times.

I guess that is another way to solve the problem where the whole deployment failed in the middle. Seems to be hardhat-deploy's philosophy:

```
In any case, as a general advice every deploy function should be idempotent. This is so they can always recover from failure or pending transaction. This is what underpin most of hardhat-deploy philosophy.
```